esutil/bulk_indexer.go

```diff
@@ -0,0 +1,509 @@
+// Elasticsearch B.V. licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information.
+
+package esutil
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"runtime"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/elastic/go-elasticsearch/v8"
+	"github.com/elastic/go-elasticsearch/v8/esapi"
+)
+
+// BulkIndexer represents a parallel, asynchronous, efficient indexer for Elasticsearch.
+//
+type BulkIndexer interface {
+	// Add adds an item to the indexer. It returns an error when the item cannot be added.
+	// Use the OnSuccess and OnFailure callbacks to get the operation result for the item.
+	//
+	// You must call the Close() method after you're done adding items.
+	//
+	// It is safe for concurrent use. When it's called from goroutines,
+	// they must finish before the call to Close, eg. using sync.WaitGroup.
+	Add(context.Context, BulkIndexerItem) error
+
+	// Close waits until all added items are flushed and closes the indexer.
+	Close(context.Context) error
+
+	// Stats returns indexer statistics.
+	Stats() BulkIndexerStats
+}
+
+// BulkIndexerConfig represents configuration of the indexer.
+//
+type BulkIndexerConfig struct {
+	NumWorkers    int           // The number of workers. Defaults to runtime.NumCPU().
+	FlushBytes    int           // The flush threshold in bytes. Defaults to 5MB.
+	FlushInterval time.Duration // The flush threshold as duration. Defaults to 10sec.
+
+	Client      *elasticsearch.Client   // The Elasticsearch client.
+	Decoder     BulkResponseJSONDecoder // A custom JSON decoder.
+	DebugLogger BulkIndexerDebugLogger  // An optional logger for debugging.
+
+	// Parameters of the Bulk API.
+	Index               string
+	ErrorTrace          bool
+	FilterPath          []string
+	Header              http.Header
+	Human               bool
+	Pipeline            string
+	Pretty              bool
+	Refresh             string
+	Routing             string
+	Source              []string
+	SourceExcludes      []string
+	SourceIncludes      []string
+	Timeout             time.Duration
+	WaitForActiveShards string
+}
+
+// BulkIndexerStats represents the indexer statistics.
+//
+type BulkIndexerStats struct {
+	NumAdded    uint
+	NumFlushed  uint
+	NumFailed   uint
+	NumIndexed  uint
+	NumCreated  uint
+	NumUpdated  uint
+	NumDeleted  uint
+	NumRequests uint
+}
+
+// BulkIndexerItem represents an indexer item.
+//
+type BulkIndexerItem struct {
+	Index           string
+	Action          string
+	DocumentID      string
+	Body            io.Reader
+	RetryOnConflict *int
+
+	Metadata interface{} // To use eg. in OnSuccess/OnFailure callbacks
+
+	OnSuccess func(item BulkIndexerItem, res BulkIndexerResponseItem)            // Per item
+	OnFailure func(item BulkIndexerItem, res BulkIndexerResponseItem, err error) // Per item
+}
+
+// BulkIndexerResponse represents the Elasticsearch response.
+//
+type BulkIndexerResponse struct {
```

Comment:
Another pro for removing configurable decoder: this type could be unexported/removed, making it possible to decode "items" in a streaming fashion (not necessarily immediately, but at some point).

Reply:
Yes, correct — thanks for the explanation.